File: exercise/ts/day15/tests/business.spec.ts
import {Factory, Inventory, WishList} from "../src/dependencies";
import {Child, Gift, ManufacturedGift} from "../src/models";
import {Business} from "../src/business";

describe('Business Logic', () => {
    let factory: Factory;
    let inventory: Inventory;
    let wishList: WishList;
    let john: Child;
    let toy: Gift;
    let manufacturedGift: ManufacturedGift;

    beforeEach(() => {
        factory = new Factory();
        inventory = new Inventory();
        wishList = new WishList();

        john = new Child('John');
        toy = new Gift('Toy');
        manufacturedGift = new ManufacturedGift('123');
    });

    test('Gift should be loaded into Sleigh', () => {
        wishList.set(john, toy);
        factory.set(toy, manufacturedGift);
        inventory.set('123', toy);

        const business = new Business(factory, inventory, wishList);
        const sleigh = business.loadGiftsInSleigh(john);

        expect(sleigh.get(john)).toBe('Gift: Toy has been loaded!');
    });

    test('Gift should not be loaded if child is not on the wish list', () => {
        const business = new Business(factory, inventory, wishList);
        const sleigh = business.loadGiftsInSleigh(john);

        expect(sleigh.has(john)).toBe(false);
    });

    test('Gift should not be loaded if the toy was not manufactured', () => {
        wishList.set(john, toy);

        const business = new Business(factory, inventory, wishList);
        const sleigh = business.loadGiftsInSleigh(john);

        expect(sleigh.has(john)).toBe(false);
    });

    test('Gift should not be loaded if the toy was misplaced', () => {
        wishList.set(john, toy);
        factory.set(toy, manufacturedGift);

        const business = new Business(factory, inventory, wishList);
        const sleigh = business.loadGiftsInSleigh(john);

        expect(sleigh.has(john)).toBe(false);
    });
});

File: exercise/scala/day15/src/test/scala/BusinessSpec.scala
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers
import santaChristmasList.operations.*

val factory = new Factory()
val inventory = new Inventory()
val wishList = new WishList()

val john: Child = Child("John")
val toy: Gift = Gift("Toy")
val manufacturedGift: ManufacturedGift = ManufacturedGift("123")

class BusinessSpec extends AnyFunSuite with Matchers {

  test("Gift should be loaded into Sleigh") {
    wishList.put(john, toy)
    factory.put(toy, manufacturedGift)
    inventory.put("123", toy)

    val business = new Business(factory, inventory, wishList)
    val sleigh = business.loadGiftsInSleigh(john)

    sleigh(john) should be("Gift: Toy has been loaded!")
  }

  test("Gift should not be loaded if child is not on the wish list") {
    val business = new Business(factory, inventory, wishList)
    val sleigh = business.loadGiftsInSleigh(john)

    sleigh should not contain john
  }

  test("Gift should not be loaded if the toy was not manufactured") {
    wishList.put(john, toy)

    val business = new Business(factory, inventory, wishList)
    val sleigh = business.loadGiftsInSleigh(john)

    sleigh should not contain john
  }

  test("Gift should not be loaded if the toy was misplaced") {
    wishList.put(john, toy)
    factory.put(toy, manufacturedGift)

    val business = new Business(factory, inventory, wishList)
    val sleigh = business.loadGiftsInSleigh(john)

    sleigh should not contain john
  }
}
